src/Microsoft.Extensions.FileProviders.Physical/ConcurrentLruCache.cs

```diff
@@ -0,0 +1,247 @@
+            _nodeList.AddFirst(node);
+        }
+
+        /// <summary>
+        /// Doesn't lock.
+        /// </summary>
+        private void UnsafeAdd(K key, V value, bool throwExceptionIfKeyExists)
+        {
+            CacheValue result;
+            if (_cache.TryGetValue(key, out result))
+            {
+                if (throwExceptionIfKeyExists)
+                {
+                    throw new ArgumentException("Key already exists", nameof(key));
+                }
+                else if (!result.Value.Equals(value))
+                {
+                    result.Value = value;
+                    _cache[key] = result;
+                    MoveNodeToTop(result.Node);
+                }
+            }
+            else
+            {
+                if (_cache.Count == _capacity)
+                {
+                    UnsafeEvictLastNode();
+                }
+                UnsafeAddNodeToTop(key, value);
+            }
+        }
+
+        public V this[K key]
+        {
+            get
+            {
+                V value;
+                if (TryGetValue(key, out value))
+                {
+                    return value;
+                }
+                else
+                {
+                    throw new KeyNotFoundException();
+                }
+            }
+            set
+            {
+                lock (_lockObject)
+                {
+                    UnsafeAdd(key, value, false);
+                }
+            }
+        }
+
+        public bool TryGetValue(K key, out V value)
+        {
+            lock (_lockObject)
+            {
+                return UnsafeTryGetValue(key, out value);
+            }
+        }
+
+        /// <summary>
+        /// Doesn't lock.
+        /// </summary>
+        public bool UnsafeTryGetValue(K key, out V value)
+        {
+            CacheValue result;
+            if (_cache.TryGetValue(key, out result))
+            {
+                MoveNodeToTop(result.Node);
+                value = result.Value;
+                return true;
+            }
+            else
+            {
+                value = default(V);
+                return false;
+            }
+        }
+
+        public V GetOrAdd(K key, V value)
+        {
+            lock (_lockObject)
+            {
+                V result;
+                if (UnsafeTryGetValue(key, out result))
+                {
+                    return result;
+                }
+                else
+                {
+                    UnsafeAdd(key, value, true);
+                    return value;
+                }
+            }
+        }
+
+        public V GetOrAdd(K key, Func<V> creator)
+        {
+            lock (_lockObject)
+            {
+                V result;
+                if (UnsafeTryGetValue(key, out result))
+                {
+                    return result;
+                }
+                else
+                {
+                    var value = creator();
```

Comment:
`Creator()` is called inside the Lock. So the File Provider is virtually freezed as long as it has not the information. You don't think it will create a bottleneck in case of simultaneous requests?